### PR TITLE
add audio support

### PR DESCRIPTION
**Pull request by marianna13**
Now it's possible to download not only video but audio as well as resample the audio (currently supports `output="files"`). Audio is being saved as `.mp3` file.

### CODE REVIEW

File: video2dataset/video2dataset.py [hunk 1]
```diff
@@ -7,51 +7,66 @@
...
 def video2dataset(
-  src,
-  dest="",
-  output_format="webdataset",
-  metadata_columns="",
+    src,
+    dest="",
+    output_format="webdataset",
+    metadata_columns="",
+    get_audio=False,
```
Comment (iejMac) [inline video2dataset/video2dataset.py:14]: can we change this to the "video_formats" parameter we discussed? This would be a comma separated string like "mp4" or "mp4,mp3" and if you want audio you can just check for "mp3" in video_formats

File: video2dataset/downloader.py [hunk 2]
```diff
@@ -24,19 +25,31 @@ def handle_youtube(youtube_url):
...
+        resp = requests.get(mp4_link, stream=True)
+        ntf = tempfile.NamedTemporaryFile()  # pylint: disable=consider-using-with
+        ntf.write(resp.content)
+        ntf.seek(0)
+        dst_name = mp4_link.split("/")[-1][:-4] + ".npy"
+        return ntf.name, ntf, None, dst_name
 
 
-def handle_url(url):
+def handle_url(url, sample_rate=None, get_audio=False):
```
Comment (iejMac) [inline video2dataset/downloader.py:47]: both parameters not necessary. If sample_rate is not None then get_audio must be True, just use that

File: video2dataset/downloader.py [hunk 3]
```diff
@@ -46,15 +59,16 @@ def handle_url(url):
     if "youtube" in url:  # youtube link
         load_file, name = handle_youtube(url)
```
Comment (iejMac) [inline video2dataset/downloader.py:60]: this doesn't work with YouTube videos yet

File: video2dataset/downloader.py [hunk 4]
```diff
@@ -24,19 +25,31 @@ def handle_youtube(youtube_url):
...
-def handle_mp4_link(mp4_link):
-    resp = requests.get(mp4_link, stream=True)
-    ntf = tempfile.NamedTemporaryFile()  # pylint: disable=consider-using-with
-    ntf.write(resp.content)
-    ntf.seek(0)
-    dst_name = mp4_link.split("/")[-1][:-4] + ".npy"
-    return ntf, dst_name
+def handle_mp4_link(
+    mp4_link: str,
+    extract_audio: bool = False,
```
Comment (iejMac) [inline video2dataset/downloader.py:30]: just use ```sample_rate is not None```

File: video2dataset/downloader.py [hunk 5]
```diff
@@ -24,19 +25,31 @@ def handle_youtube(youtube_url):
...
+    sample_rate: int = None
+):
+    if extract_audio:
```
Comment (iejMac) [inline video2dataset/downloader.py:33]: you should utilize the video bytes for this and also give the option to return both as we discussed in a dict with the extensions as the keys

File: video2dataset/video2dataset.py [hunk 6]
```diff
@@ -7,51 +7,66 @@
...
+        for k in meta:
+            vid_meta[k] = meta[k][i].as_py()
+
+        # NOTE: Right now assuming video is url (maybe add support for local mp4
+        load_vid, file, dst_name, info = handle_url(
+            vid, get_audio=get_audio, sample_rate=sample_rate)
+        if info:
+            vid_meta = {
+                'meta': vid_meta,
+                'audio_info': info
```
Comment (iejMac) [inline video2dataset/video2dataset.py:60]: why do you need this at all?

File: video2dataset/video2dataset.py (hunk 1 above)
Comment (marianna13) [reply video2dataset/video2dataset.py:14]: Already implemented, working on testing

File: video2dataset/downloader.py (hunk 2 above)
Comment (marianna13) [reply video2dataset/downloader.py:47]: I removed get_audio argument

File: video2dataset/downloader.py (hunk 4 above)
Comment (marianna13) [reply video2dataset/downloader.py:30]: I will use video_format argument

File: video2dataset/downloader.py (hunk 5 above)
Comment (marianna13) [reply video2dataset/downloader.py:33]: Already implemented, testing now

File: video2dataset/video2dataset.py (hunk 6 above)
Comment (marianna13) [reply video2dataset/video2dataset.py:60]: I got rid of it, you're right I don't need it

File: video2dataset/audio_utils.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
+import json
```
Comment (rom1504) [inline video2dataset/audio_utils.py:1]: let's not have an utils file, and move this into downloader

Comment (marianna13) [reply video2dataset/audio_utils.py]: ok

File: video2dataset/reader.py [hunk 8]
```diff
@@ -17,7 +17,7 @@ class Reader:
     anything else - put in key.json metadata file
     """
 
-    def __init__(self, src, meta_columns=None):
+    def __init__(self, src, meta_columns=None, url_col='videoLoc'):
```
Comment (iejMac) [inline video2dataset/reader.py:20]: what is this? why is this in the audio PR?

File: video2dataset/video2dataset.py [hunk 9]
```diff
@@ -4,54 +4,63 @@
...
-  dest="",
-  output_format="webdataset",
-  metadata_columns="",
+    src,
+    dest="",
+    output_format="webdataset",
+    metadata_columns="",
+    video_format="",
+    sample_rate=None,
+    url_col='videoLoc'
```
Comment (iejMac) [inline video2dataset/video2dataset.py:17]: why do we have this PR, why are we making unnecessary changes in the audio PR

**Comment by iejMac** [comment]
@marianna13 please let's try to keep this as small as possible, no unnecessary changes to other components, if you think something else should be changed let's discuss over discord and then make a different PR. If this PR is about "audio support" then the changes in the diff should reflect integrating audio support and not refactoring how we access the video link

**Comment by rom1504** [comment]
I changed things significantly so I'm afraid this will need to be redone

could you try to implement these features:
* reading local mp3 files
* reading audio from yt-dlp 

it would be great!

**Comment by marianna13** [comment]
Sure!
1. Does it need to support only mp3 or other audio formats as well?
2. What do you mean by "reading audio from yt-dlp"? I can extract audio link (m4a or mp3 or whatever) and download audio from it or I could just download audio with yt-dlp.

**Comment by rom1504** [comment]
all sounds great, but I figure start with something minimal and then we can expand on top

**Comment by iejMac** [comment]
@marianna13 remember to merge in main

File: tests/test_audio.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+import pytest
+import tempfile
+import subprocess
+from video2dataset.data_reader import VideoDataReader
+import json
+
+
+@pytest.mark.parametrize("input_file", ["test_webvid.csv", "test_yt.csv"])
+def test_audio(input_file):
+    encode_formats = {"video": "mp4", "audio": "mp3", "sample_rate": 16000}
```
Comment (iejMac) [inline tests/test_audio.py:13]: can you test out a few other sample_rates?

File: tests/test_audio.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+        for i, url in enumerate(url_list):
+            key, vid_bytes, aud_bytes, yt_meta_dict, error_message = video_data_reader((i, url))
+            if aud_bytes is not None:
+                with tempfile.NamedTemporaryFile(suffix=".mp3") as f:
+                    f.write(aud_bytes)
+
+                    out = subprocess.check_output(f"file {f.name}".split()).decode("utf-8")
+                    assert "Audio file with ID3 version" in out
+
+                    result = subprocess.check_output(
```
Comment (iejMac) [inline tests/test_audio.py:29]: can you use ffmpeg-python API like I do in other tests?

File: video2dataset/data_reader.py [hunk 12]
```diff
@@ -5,6 +5,20 @@
 import yt_dlp
 import io
 import webvtt
+import subprocess
+
+
+def video2audio(video, af, sample_rate, tmp_dir):
+    """extract video from audio"""
```
Comment (iejMac) [inline video2dataset/data_reader.py:12]: audio from video

File: video2dataset/data_reader.py [hunk 13]
```diff
@@ -5,6 +5,20 @@
...
+    path = f"{tmp_dir}/{str(uuid.uuid4())}.{af}"
+    sr_str = f"-ar {sample_rate}" if sample_rate else ""
+    cmd = f"ffprobe -v error -show_entries format=nb_streams -of default=noprint_wrappers=1:nokey=1 {video}"
+    num_streams = subprocess.check_output(cmd.split())
+    if int(num_streams.decode("utf-8")) == 2:
+        os.system(f"ffmpeg -v error -i {video} -vn -ac 2 {sr_str} -ab 320k -f {af} {path}")
```
Comment (iejMac) [inline video2dataset/data_reader.py:18]: can you use FFmpeg python API like we do everywhere else?

File: video2dataset/data_reader.py [hunk 14]
```diff
@@ -123,37 +150,52 @@ def __call__(self, url):
...
+                    vid_bytes = vid_file.read()
+            if a_file_path is not None:
+                with open(a_file_path, "rb") as aud_file:
+                    aud_bytes = aud_file.read()
         else:
             vid_bytes = None
         if file_path is not None:  # manually remove tempfile
             os.remove(file_path)
-        return key, vid_bytes, yt_meta_dict, error_message
+        return key, vid_bytes, aud_bytes, yt_meta_dict, error_message
```
Comment (iejMac) [inline video2dataset/data_reader.py:201]: can we return dict instead of video and audio bytes separately. Just make a dict with same keys as encode_formats but value is the bytes

File: video2dataset/worker.py [hunk 15]
```diff
@@ -159,12 +155,19 @@ def data_generator():
                         semaphore.release()
                         continue
 
-                    bytes_downloaded += len(vid_stream)
+                    if vid_stream is not None:
+                        bytes_downloaded += len(vid_stream)
+                    if aud_stream is not None:
+                        bytes_downloaded += len(aud_stream)
```
Comment (iejMac) [inline video2dataset/worker.py:161]: not sure this is correct if you extract the audio from the video that means the bytes of the audio are in the video. don't duplicate them

File: video2dataset/worker.py [hunk 16]
```diff
@@ -199,6 +203,15 @@ def data_generator():
                             meta["key"],
                             sample_data[caption_indice] if caption_indice is not None else None,
                             meta,
+                            format_type="video",
+                        )
+                    if aud_stream is not None:
```
Comment (iejMac) [inline video2dataset/worker.py:208]: instead of doing it like this can't we just pass streams into here the sample writer? The sample writer should know all the streams to expect because you pass in encode_formats to it so can't we just write those all together?

**Comment by rom1504** [comment]
does this work fast?

**Comment by marianna13** [comment]
Speed test (1 video, on CPU cluster):
- only video (mp4): 8.223158836364746 seconds
- only audio (mp3): 8.929118394851685 seconds
- video + audio (mp4, mp3): 8.74603796005249 seconds

**Comment by rom1504** [comment]
Lgtm

**Comment by rom1504** [comment]
There's a few typos and bad variable names but not a big deal
